README.md
# node-libravatar

A library built for easy integration with the federated [Libravatar](http://www.libravatar.org)
avatar hosting service from within your Node.js applications. The project was inspired by a 
[Gravatar library](https://github.com/emerleite/node-gravatar) from [Emerson Macedo](http://codificando.com/).

See the [project page](https://github.com/coloradocolby/node-libravatar) for the issue tracker and downloads.

[![Build Status](https://travis-ci.org/coloradocolby/node-libravatar.png)](https://travis-ci.org/coloradocolby/node-libravatar)
[![Dependencies Status](https://david-dm.org/coloradocolby/node-libravatar.png)](https://david-dm.org/coloradocolby/node-libravatar)
[![Dev Dependencies Status](https://david-dm.org/coloradocolby/node-libravatar/dev-status.png)](https://david-dm.org/coloradocolby/node-libravatar#info=devDependencies)
[![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/coloradocolby/node-libravatar/blob/master/LICENSE.md)

## Installation

To install using npm:

    $ npm install libravatar

[![NPM](https://nodei.co/npm/libravatar.png)](https://nodei.co/npm/libravatar/)

## Usage

To generate the correct avatar URL based on someone's email address, use the
following:

    const libravatar = require('libravatar')
    ...
    const getLibravatar = async () => {
      const avatar_url = await libravatar.get_avatar_url({ email: 'person@example.com', size: 96, default: 'mm', https: false })
    }

See the [Libravatar documentation](http://wiki.libravatar.org/api) for more
information on the special values for the "default" parameter.


CHANGELOG.md
**3.0.0 (2019-09-07):**
- Async/ES6 integration

**2.0.0 (2012-04-22):**
- Backwards-incompatible API change for the url() function
- Minor style improvements in the code

**1.1.1 (2012-03-02):**
- Minor code cleanups from Andy Chilton
- Add missing dev dependency on tap

**1.1.0 (2011-11-20):**
- Fix federation if more than one SRV record is present
- Fix bugs in parameter validation and URL normalization
- Add unit tests

**1.0.1 (2011-10-17):**
- Add an error param to the callback
- Style improvements

**1.0.0 (2011-10-16):**
- Initial public release


tests/url.js
const test = require('tap').test

const libravatar = require('../lib/libravatar')

const COMMON_EMAIL = 'whatever@wherever.whichever'
const COMMON_EMAIL_HASH = 'a60fc0828e808b9a6a9d50f1792240c8'
const COMMON_EMAIL_DOMAIN = 'wherever.whichever'
const COMMON_OPENID = 'http://example.com/id'
const COMMON_OPENID_HASH =
  'ce0064bb30c22b618f814c389e7941ce1bfff0659910523192868d2b71632c77'
const COMMON_OPENID_DOMAIN = 'example.com'
const COMMON_PREFIX_HTTP = 'http://cdn.libravatar.org/avatar/'
const COMMON_PREFIX_HTTPS = 'https://seccdn.libravatar.org/avatar/'

const array_equal = (t, result, expected, description) => {
  t.equal(result[0], expected[0], description + ' [0]')
  t.equal(result[1], expected[1], description + ' [1]')
}

test('parsing of user identity', t => {
  const test1 = libravatar.parse_user_identity(null, null)
  const exp1 = [null, null]
  array_equal(t, test1, exp1, 'both parameters missing')

  const test2 = libravatar.parse_user_identity(COMMON_EMAIL, COMMON_OPENID)
  const exp2 = [COMMON_EMAIL_HASH, COMMON_EMAIL_DOMAIN]
  array_equal(t, test2, exp2, 'both parameters supplied')

  const test3 = libravatar.parse_user_identity(COMMON_EMAIL, null)
  const exp3 = [COMMON_EMAIL_HASH, COMMON_EMAIL_DOMAIN]
  array_equal(t, test3, exp3, 'standard email')

  const test4 = libravatar.parse_user_identity(null, COMMON_OPENID)
  const exp4 = [COMMON_OPENID_HASH, COMMON_OPENID_DOMAIN]
  array_equal(t, test4, exp4, 'standard openid')

  t.end()
})

test('parsing of email address', t => {
  const test1 = libravatar.parse_user_identity('', null)
  const exp1 = [null, null]
  array_equal(t, test1, exp1, 'empty email')

  const test2 = libravatar.parse_user_identity('username', null)
  const exp2 = [null, null]
  array_equal(t, test2, exp2, 'missing hostname')

  const test3 = libravatar.parse_user_identity(
    'WHATEVER@wherever.whichever',
    null
  )
  const exp3 = [COMMON_EMAIL_HASH, COMMON_EMAIL_DOMAIN]
  array_equal(t, test3, exp3, 'uppercase username')

  const test4 = libravatar.parse_user_identity(
    'Whatever@Wherever.whichever',
    null
  )
  const exp4 = [COMMON_EMAIL_HASH, COMMON_EMAIL_DOMAIN]
  array_equal(t, test4, exp4, 'mixed-case username and hostname')

  const test5 = libravatar.parse_user_identity(
    ' Whatever@Wherever.whichever   ',
    null
  )
  const exp5 = [COMMON_EMAIL_HASH, COMMON_EMAIL_DOMAIN]
  array_equal(t, test5, exp5, 'untrimmed email')

  t.end()
})

test('parsing of openid urls', t => {
  const test1 = libravatar.parse_user_identity(null, '')
  const exp1 = [null, null]
  array_equal(t, test1, exp1, 'empty openid')

  const test2 = libravatar.parse_user_identity(null, 'url')
  const exp2 = [null, null]
  array_equal(t, test2, exp2, 'invalid url')

  const test3 = libravatar.parse_user_identity(null, 'http://example.COM/id')
  const exp3 = [COMMON_OPENID_HASH, COMMON_OPENID_DOMAIN]
  array_equal(t, test3, exp3, 'mixed-case hostname')

  const test4 = libravatar.parse_user_identity(
    null,
    '  HTTP://example.com/id  '
  )
  const exp4 = [COMMON_OPENID_HASH, COMMON_OPENID_DOMAIN]
  array_equal(t, test4, exp4, 'uppercase scheme')

  const test5 = libravatar.parse_user_identity(
    null,
    'http://user:password@Example.com/id'
  )
  const exp5 = [
    'e1cf8061371aa00b82c0cf0b9b1140546bc31cd4a15cb8adc84ad01823bdf71e',
    COMMON_OPENID_DOMAIN,
  ]
  array_equal(t, test5, exp5, 'lowercase basic auth')

  const test6 = libravatar.parse_user_identity(
    null,
    'http://User:Password@Example.com/id'
  )
  const exp6 = [
    '50f60bb4c1b47fffdd6e2ce65f8bf37b65a2fb960596fa6789ef7b0044b931a2',
    COMMON_OPENID_DOMAIN,
  ]
  array_equal(t, test6, exp6, 'mixed-case basic auth')

  const test7 = libravatar.parse_user_identity(
    null,
    'http://openid.example.COM/id'
  )
  const exp7 = [
    'a108913053c4949f18d9eef7a4a68f27591297cdd7a7e2e375702aa87b6d3c05',
    'openid.example.com',
  ]
  array_equal(t, test7, exp7, 'sub-domain')

  const test8 = libravatar.parse_user_identity(null, 'https://example.com/id')
  const exp8 = [
    '43e813cfff429662436728ef4fb1cc12bcf20414cab78811137f7d718c1ddedb',
    COMMON_OPENID_DOMAIN,
  ]
  array_equal(t, test8, exp8, 'https')

  const test9 = libravatar.parse_user_identity(null, 'http://example.com/ID')
  const exp9 = [
    'ad8ce775cc12cba9bb8af26e00f55c473a3fcd3f554595a5ad9dd924a546a448',
    COMMON_OPENID_DOMAIN,
  ]
  array_equal(t, test9, exp9, 'uppercase path')

  t.end()
})

test('parsing of openid urls', t => {
  const test1 = libravatar.compose_avatar_url('', '', '', false)
  const exp1 = COMMON_PREFIX_HTTP
  t.equal(test1, exp1, 'degenerate http case')

  const test2 = libravatar.compose_avatar_url('', '', '', true)
  const exp2 = COMMON_PREFIX_HTTPS
  t.equal(test2, exp2, 'degenerate https case')

  const test3 = libravatar.compose_avatar_url('', 'deadbeef', '', false)
  const exp3 = COMMON_PREFIX_HTTP + 'deadbeef'
  t.equal(test3, exp3, 'simple http hash')

  const test4 = libravatar.compose_avatar_url(
    'avatar.example.com',
    'deadbeef',
    '',
    false
  )
  const exp4 = 'http://avatar.example.com/avatar/deadbeef'
  t.equal(test4, exp4, 'federated http hash')

  const test5 = libravatar.compose_avatar_url(
    'avatar.example.com',
    'deadbeef',
    '?s=24',
    true
  )
  const exp5 = 'https://avatar.example.com/avatar/deadbeef?s=24'
  t.equal(test5, exp5, 'federated https hash with size')

  const test6 = libravatar.compose_avatar_url(
    '',
    '12345678901234567890123456789012',
    '?d=404',
    true
  )
  const exp6 = COMMON_PREFIX_HTTPS + '12345678901234567890123456789012?d=404'
  t.equal(test6, exp6, 'common https hash with default')

  t.end()
})


tests/federation.js
const test = require('tap').test
const libravatar = require('../lib/libravatar')

test('sanitization of invalid SRV responses', t => {
  const test1 = libravatar.sanitized_target([null, null], false)
  const exp1 = null
  t.equal(test1, exp1, 'both parameters missing')

  const test2 = libravatar.sanitized_target([null, 80], false)
  const exp2 = null
  t.equal(test2, exp2, 'first parameter missing')

  const test3 = libravatar.sanitized_target(['example.com', null], true)
  const exp3 = null
  t.equal(test3, exp3, 'second parameter missing')

  const test4 = libravatar.sanitized_target(['example.com', 0], false)
  const exp4 = null
  t.equal(test4, exp4, 'port too small')

  const test5 = libravatar.sanitized_target(['example.com', 70000], true)
  const exp5 = null
  t.equal(test5, exp5, 'port too big')

  const test6 = libravatar.sanitized_target(['exam$ple.com', 80], false)
  const exp6 = null
  t.equal(test6, exp6, 'invalid hostname')

  const test7 = libravatar.sanitized_target(['example.com', 'abc'], true)
  const exp7 = null
  t.equal(test7, exp7, 'invalid port')

  t.end()
})

test('sanitization of valid SRV responses', t => {
  const test1 = libravatar.sanitized_target(['example.com', 80], false)
  const exp1 = 'example.com'
  t.equal(test1, exp1, 'normal http')

  const test2 = libravatar.sanitized_target(['example.com', 443], true)
  const exp2 = 'example.com'
  t.equal(test2, exp2, 'normal https')

  const test3 = libravatar.sanitized_target(['example.org', 8080], false)
  const exp3 = 'example.org:8080'
  t.equal(test3, exp3, 'weird http')

  const test4 = libravatar.sanitized_target(['example.org', 44300], true)
  const exp4 = 'example.org:44300'
  t.equal(test4, exp4, 'weird https')

  t.end()
})

const array_equal = (t, result, expected, description) => {
  t.equal(result[0], expected[0], description + ' [0]')
  t.equal(result[1], expected[1], description + ' [1]')
}

test('ordering of invalid SRV hostnames', t => {
  const test1 = libravatar.srv_hostname([])
  const exp1 = [null, null]
  array_equal(t, test1, exp1, 'empty array')

  const test2 = libravatar.srv_hostname([{ name: null, port: null }])
  const exp2 = [null, null]
  array_equal(t, test2, exp2, 'single empty array')

  t.end()
})

test('ordering of valid SRV hostnames on priority', t => {
  const test1 = libravatar.srv_hostname([{ name: 'example.com', port: 81 }])
  const exp1 = ['example.com', 81]
  array_equal(t, test1, exp1, 'single hostname')

  const test2 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 0, weight: 0 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 0 },
  ])
  const exp2 = ['a.example.org', 81]
  array_equal(t, test2, exp2, 'two hostnames')

  const test3 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 0 },
    { name: 'b.example.org', port: 82, priority: 1, weight: 0 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 0 },
    { name: 'd.example.org', port: 84, priority: 10, weight: 0 },
  ])
  const exp3 = ['b.example.org', 82]
  array_equal(t, test3, exp3, 'four hostnames')

  t.end()
})

test('ordering of valid SRV hostnames on weight', t => {
  const original_random = Math.random

  Math.random = () => {
    return 0.6
  }
  const test1 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 1 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 5 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 10 },
    { name: 'd.example.org', port: 84, priority: 10, weight: 50 },
    { name: 'e.example.org', port: 85, priority: 10, weight: 0 },
  ])
  const exp1 = ['d.example.org', 84]
  array_equal(t, test1, exp1, 'random 1')

  Math.random = () => {
    return 0.2
  }
  const test2 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 40 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 0 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 0 },
    { name: 'e.example.org', port: 85, priority: 10, weight: 0 },
  ])
  const exp2 = ['a.example.org', 81]
  array_equal(t, test2, exp2, 'random 2')

  Math.random = () => {
    return 0.4
  }
  const test3 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 1 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 0 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 0 },
    { name: 'e.example.org', port: 85, priority: 10, weight: 0 },
  ])
  const exp3 = ['e.example.org', 85]
  array_equal(t, test3, exp3, 'random 3')

  Math.random = () => {
    return 0.3
  }
  const test4 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 0 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 0 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 10 },
    { name: 'e.example.org', port: 85, priority: 10, weight: 0 },
  ])
  const exp4 = ['c.example.org', 83]
  array_equal(t, test4, exp4, 'random 4')

  Math.random = () => {
    return 0.8
  }
  const test5 = libravatar.srv_hostname([
    { name: 'a.example.org', port: 81, priority: 10, weight: 1 },
    { name: 'b.example.org', port: 82, priority: 10, weight: 5 },
    { name: 'c.example.org', port: 83, priority: 10, weight: 10 },
    { name: 'd.example.org', port: 84, priority: 10, weight: 30 },
    { name: 'e.example.org', port: 85, priority: 10, weight: 50 },
    { name: 'f.example.org', port: 86, priority: 10, weight: 0 },
  ])
  const exp5 = ['e.example.org', 85]
  array_equal(t, test5, exp5, 'random 5')

  Math.random = original_random
  t.end()
})

test('ordering of valid SRV hostnames on weight', t => {
  const test1 = libravatar.service_name(null, false)
  const exp1 = null
  t.equal(test1, exp1, 'degenerate case')

  const test2 = libravatar.service_name('example.com', false)
  const exp2 = '_avatars._tcp.example.com'
  t.equal(test2, exp2, 'simple domain over http')

  const test3 = libravatar.service_name('example.org', true)
  const exp3 = '_avatars-sec._tcp.example.org'
  t.equal(test3, exp3, 'simple domain over https')

  const test4 = libravatar.service_name('example.co.nz', false)
  const exp4 = '_avatars._tcp.example.co.nz'
  t.equal(test4, exp4, 'longer domain over http')

  t.end()
})


lib/libravatar.js
/*
 * node-libravatar: node.js module for Libravatar
 *
 * Copyright (C) 2011, 2012 Francois Marier <francois@libravatar.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE."
 */

const crypto = require('crypto')
const dns = require('dns')
const querystring = require('querystring')
const url = require('url')

const BASE_URL = 'http://cdn.libravatar.org/avatar/'
const SECURE_BASE_URL = 'https://seccdn.libravatar.org/avatar/'
const SERVICE_BASE = '_avatars._tcp'
const SECURE_SERVICE_BASE = '_avatars-sec._tcp'

/*
 * Return the right (target, port) pair from a list of SRV records.
 */
const srv_hostname = records => {
  if (records.length < 1) {
    return [null, null]
  }

  if (1 === records.length) {
    return [records[0]['name'], records[0]['port']]
  }

  // Keep only the servers in the top priority
  let priority_records = []
  let total_weight = 0
  let top_priority = records[0]['priority'] // highest priority = lowest number

  records.forEach(srv_record => {
    if (srv_record['priority'] <= top_priority) {
      // ignore lower priority records
      if (srv_record['priority'] < top_priority) {
        // reset the array (srv_record has higher priority)
        top_priority = srv_record['priority']
        total_weight = 0
        priority_records = []
      }

      total_weight += srv_record['weight']

      if (srv_record['weight'] > 0) {
        priority_records.push([total_weight, srv_record])
      } else {
        // zero-weigth elements must come first
        priority_records.unshift([0, srv_record])
      }
    }
  })

  if (1 === priority_records.length) {
    const srv_record = priority_records[0][1]
    return [srv_record['name'], srv_record['port']]
  }

  // Select first record according to RFC2782 weight
  // ordering algorithm (page 3)
  const random_number = Math.floor(Math.random() * (total_weight + 1))

  for (let i = 0; i < priority_records.length; i++) {
    const weighted_index = priority_records[i][0]
    const target = priority_records[i][1]

    if (weighted_index >= random_number) {
      return [target['name'], target['port']]
    }
  }

  dns.resolveSrv()
  console.log('There is something wrong with our SRV weight ordering algorithm')
  return [null, null]
}

/*
 * Ensure we are getting a (mostly) valid hostname and port number
 * from the DNS resolver and return the final hostname:port string.
 */
const sanitized_target = (target_components, https) => {
  const target = target_components[0]
  const port = parseInt(target_components[1])

  if (target === null || isNaN(port)) {
    return null
  }

  if (port < 1 || port > 65535) {
    return null
  }

  if (target.search(/^[0-9a-zA-Z\-.]+$/) === -1) {
    return null
  }

  if (target && ((https && port != 443) || (!https && port != 80))) {
    return target + ':' + port
  } else {
    return target
  }
}

/*
 * Generate user hash based on the email address or OpenID and return
 * it along with the relevant domain.
 */
const parse_user_identity = (email, openid) => {
  let hash = null,
    domain = null

  if (email != null) {
    const lowercase_value = email.trim().toLowerCase()
    const email_parts = lowercase_value.split('@')
    if (email_parts.length > 1) {
      domain = email_parts[email_parts.length - 1]
      hash = crypto
        .createHash('md5')
        .update(lowercase_value)
        .digest('hex')
    }
  } else if (openid != null) {
    const parsed_url = url.parse(openid)
    if (parsed_url.protocol && parsed_url.hostname) {
      let normalized_url = parsed_url.protocol.toLowerCase()
      normalized_url += parsed_url.slashes ? '//' : ''
      if (parsed_url.auth) {
        normalized_url += parsed_url.auth + '@'
      }
      normalized_url += parsed_url.hostname.toLowerCase() + parsed_url.pathname

      domain = parsed_url.hostname.toLowerCase()
      hash = crypto
        .createHash('sha256')
        .update(normalized_url)
        .digest('hex')
    }
  }

  return [hash, domain]
}

/*
 * Return the DNS service to query for a given domain and scheme.
 */
const service_name = (domain, https) => {
  if (domain) {
    return (https ? SECURE_SERVICE_BASE : SERVICE_BASE) + '.' + domain
  }
  return null
}

/*
 * Assemble the final avatar URL based on the provided components.
 */
const compose_avatar_url = (
  delegation_server,
  avatar_hash,
  query_string,
  https
) => {
  let base_url = (https && SECURE_BASE_URL) || BASE_URL

  if (delegation_server) {
    base_url = `http${https ? 's' : ''}://${delegation_server}/avatar/`
  }

  return base_url + avatar_hash + query_string
}

const get_delegation_server = (domain, https) => {
  return new Promise(resolve => {
    dns.resolveSrv(service_name(domain, https), (err, addresses) => {
      let delegation_server
      if (err === null) {
        delegation_server = sanitized_target(srv_hostname(addresses), https)
      }
      resolve(delegation_server)
    })
  })
}

const get_avatar_url = async options => {
  const identity = parse_user_identity(options.email, options.openid)
  const hash = identity[0]
  const domain = identity[1]
  const https = options.https || false
  if (hash) {
    delete options.email
    delete options.openid
    delete options.https
    const query_data = querystring.stringify(options)
    const query = (query_data && '?' + query_data) || ''

    const delegation_server = await get_delegation_server(domain, https)

    return compose_avatar_url(delegation_server, hash, query, https)
  } else {
    throw new Error('An email or an OpenID must be provided.')
  }
}

const libravatar = {
  // These ones are exported for the unit tests only
  sanitized_target,
  srv_hostname,
  parse_user_identity,
  service_name,
  compose_avatar_url,
  get_avatar_url,
}

module.exports = libravatar
